Ghgc 632/scheduled worldview nrt collection update

sanzog03 opened the pull request:
**Summary:**

Addresses [GHGC-632: Automated STAC update for the updated Worldview NRT datasets](https://github.com/US-GHG-Center/ghgc-architecture/issues/632)

## Changes

* Create a DAG capable of indexing WMTS into STAC.
* Scheduler and gibs url are optional, but when provided, the task for scheduled update (necessary for NRT) will trigger based off gibs metadata. Collection Id should be common in STAC and GIBS.
* There is validation task added for config json to check if it adheres to the WMTS extension specification.

## PR Checklist

- [x] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

sanzog03 committed:
create a dag to handle the daily scheduled nightlight NRT data. The data update DAG runs validation task to check if the source has latest NRT data

sanzog03 committed:
update the nrt update check and nrt data update task

sanzog03 committed:
refactor to use the date extracted from metadata, into the collection update

sanzog03 committed:
replace space with underscore on id

kyle-lesinger commented on dags/veda_data_pipeline/veda_worldview_nrt_collection_update_pipeline.py:
```diff
@@ -0,0 +1,115 @@
+    "tags": ["collection"],
+}
+
+with DAG(
```
In this section, it looks like its only set for a manual trigger. If you want to include make this DAG run every day, you would need to include the `schedule` parameter.

```
    with DAG(
            id,
            schedule="0 5 * * *", # runs at 5am everyday
            render_template_as_native_obj=True,   # required to use params in the DAG
            params=template_conf,
            **dag_args
    ) as dag:
        start = EmptyOperator(task_id="Start", dag=dag)
        end = EmptyOperator(task_id="End", trigger_rule=TriggerRule.ONE_SUCCESS, dag=dag)
```

kyle-lesinger commented on dags/veda_data_pipeline/veda_worldview_nrt_collection_update_pipeline.py:
```diff
@@ -0,0 +1,115 @@
+    "tags": ["collection"],
+}
+
+with DAG(
```
In this section, it looks like its only set for a manual trigger. If you want to include make this DAG run every day, you would need to include the schedule parameter.

```
with DAG(
        id,
        schedule="0 5 * * *", # runs at 5am everyday
        render_template_as_native_obj=True,   # required to use params in the DAG
        params=template_conf,
        **dag_args
) as dag:
    start = EmptyOperator(task_id="Start", dag=dag)
    end = EmptyOperator(task_id="End", trigger_rule=TriggerRule.ONE_SUCCESS, dag=dag)
```

sanzog03 replied:
noted and changed

kyle-lesinger commented on dags/veda_data_pipeline/groups/collection_group.py:
```diff
@@ -79,3 +80,86 @@ def collection_task_group():
     generate_collection = generate_collection_task()
     ingest_collection = ingest_collection_task(collection=generate_collection)
 
+# Special task group to update nightlight NRT data collection that is pulled from worldview
```
Lines 83-165 only relate to Worldview while the other functions (lines 12-83) are general to all tasks. @amarouane-ABDELHAK would it be better to move Worldview tasks to remain within [veda_worldview_nrt_collection_update_pipeline.py](https://github.com/NASA-IMPACT/veda-data-airflow/pull/378/files#diff-151a1f805d982d9f8aaeae83d358a1b50fbca108713f69ba4923975a890cc97d)?

sanzog03 replied:
noted.

sanzog03 committed:
update code into more granular subgroups. also moved nrt update specific group code to its own concern.

sanzog03 committed:
removed unnecessary import

sanzog03 committed:
fix proper scoping on the branch_task return, enabling followup task

sanzog03 committed:
refactor dag to use taskflow api completely. added docstring to the tasks. also extracted key variables as global for reusability

sanzog03 committed:
add stac_extension to include web map links used by wmts capabilities.

sanzog03 committed:
generic def names

sanzog03 committed:
task group taking whatever is needed via parameters instead of global

sanzog03 committed:
modified dag creation with the thought of re-useability

sanzog03 committed:
generate dag docs based on the reused dag

sanzog03 committed:
add veda_worldview_nrt_data_collection_update_dag_creator to generate dag

sanzog03 committed:
changed variables names and dag id generation to be more semantic

sanzog03 committed:
change the taskgroupid to be generic

sanzog03 committed:
add relevant dag tags

sanzog03 committed:
fix collection config

slesaad commented:
We want to make a generic dag and then put our specific dag config in the s3 bucket. This basically hardcodes the collection within the dag itself. Let's talk

sanzog03 commented:
> We want to make a generic dag and then put our specific dag config in the s3 bucket. This basically hardcodes the collection within the dag itself. Let's talk

The hardcoded aspects within the DAG itself are for its default settings. We then pull the unique DAG configurations from the S3 bucket, as you mentioned.

sanzog03 committed:
added some comments

sanzog03 committed:
renamed the dags and configs for better clarity

sanzog03 committed:
Merge branch 'dev' into GHGC-632/scheduled_worldview_nrt_collection_update

sanzog03 committed:
remove unnecessary code

sanzog03 committed:
take schedule from generate dag config. better namings

sanzog03 committed:
add ability to ingest wmts without gibs

sanzog03 committed:
remove unwanted extensions from wmts config json

sanzog03 committed:
adhere to the return type

sanzog03 committed:
single task instantiation for whatever is necessary

sanzog03 committed:
refactor for better task arrangement and readability

sanzog03 committed:
import error

sanzog03 committed:
1. made gibs url and schedule optional.
2. fix to the task group call.

sanzog03 committed:
wmts and gibs branching fix

sanzog03 committed:
rename to make the dag generic for all wmts and not just gibs wmts

sanzog03 committed:
update doc string

sanzog03 committed:
1. add validation task which validates if the json config is as per the stac extension specs.
2. only necessary inputs for dag execution
3. change the config to adhere to the wmts extension specs

sanzog03 committed:
remove comment

sanzog03 committed:
corrected import for tests

kyle-lesinger commented on dags/veda_data_pipeline/helpers/veda_wmts2stac_update_pipeline.py:
```diff
@@ -0,0 +1,405 @@
+                return layer_date
+    return ""
+
+@task
```
Unsure if we need this validation config explicitly added. If we have a pystac collection, there is a method to just call `collection.validate()` and it should lookup the github repo and schema and run validation checks. See [example notebook here](https://docs.openveda.cloud/user-guide/notebooks/veda-operations/stac-collection-creation.html) for how VEDA has previously done it.

kyle-lesinger replied:
Just to leave another comment, this isn't a blocker. Maybe something to incorporate in newer DAGs which create collections.

sanzog03 replied:
The pystac validate() does not validate the extensions; There is no implementation to validate for extensions. ref https://github.com/stac-utils/pystac/blob/main/pystac/validation/stac_validator.py#L65 <img width="2022" height="1218" alt="Image" src="https://github.com/user-attachments/assets/2ba1a127-edd8-4691-adcd-18125fdb3932" />

sanzog03 replied:
We have migrated to pystac validation.

kyle-lesinger committed:
update wmts with validator

kyle-lesinger committed:
update pytests

kyle-lesinger committed:
add slack fail alert

kyle-lesinger committed:
add value error messages

kyle-lesinger committed:
reduce number of logging statements

slesaad reviewed (commented):
Looks good to me.

@anayeaye @smohiudd appreciate it if you guys could have a look too!

kyle-lesinger committed:
update generate dags
